### ISSUE TITLE
Performance compared to other simpler algorithm, and keygen improvement

### ISSUE THREAD

**Issue by ganego** [SethMMorton/natsort#110]
I use the following function to have natsort and it worked fine for all my usecases until now - now I needed to sort paths with subfolders, and your library does a very good job at sorting those correctly via `ns.PATH`, so thank you for that.

```
def sorted_alphanumeric(data):
    convert = lambda text: int(text) if text.isdigit() else text.lower()
    alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ]
    return sorted(data, key=alphanum_key)
```

For a simple test case (for non-path strings like `["a", "b", "c", "d"]`), it's 10x faster than your library. I tested this by simply running each sort 10k times.

Not sure if something can be implemented to sort out "simple" cases like those and use a faster method, but I'd like to propose another way where I could speed up your library in my test case a lot: 

The problem mainly is creating the 'key' which takes time. So if we create they key once and then use the sort 10k times, the keygen has to run only twice.  

In `def natsorted(seq, key=None, reverse=False, alg=ns.DEFAULT)` you allow `key` to be given to the function, yet giving it does absolutely nothing. You still run `key = natsort_keygen(key, alg)` every time, and the `key` will be just passed along.

So add `if not key` before `key = natsort_keygen(key, alg)`. Hence, when given an already generated key, you won't generate it again. From doing that I could speed up my test from being 13x slower to only being 2x slower.

**Comment by SethMMorton** [comment]
I'm having a hard time following exactly what the issue is, but I *think* you are saying that when you compare your custom sort method to `natsorted` then `natsorted` is slow because each time it is invoked it has to regenerate the key. This is true. If you want to minimize the time spent in generating the key, I recommend the following:

```python
>>> import natsort
>>> mykey = natsort.natsort_keygen(<my arguments>)
>>> sorted([<my data>], key=mykey)
```

One could probably implement a caching mechanism for the key so that if the same options are passed a previous key could be reused.

---

> you allow key to be given to the function, yet giving it does absolutely nothing. You still run `key = natsort_keygen(key, alg)` every time, and the key will be just passed along.

The `key argument does not do nothing. It is used to transform the input before being sent to the `natsort` algorithm. It sounds to me like you wished that passing the `key` completely bypassed the `natsort` algorithm... if this is the desired behavior, I recommend using `sorted`.

**Comment by ganego** [comment]
I kinda mixed two issues:

I) My (well from StackOverflow) function is faster, even without regenerating the key via natsort every cycle. I meant to ask, if I makes sense for natsort to implement some test to use a faster function to do natural sorting, like in the simple example, where there are only simple strings.

II) I might have not looked into it well enough or understood it well enough (not a professional coder). I saw that when using `natsorted(...)` it would create the `key` again each cycle. So I thought, 'let's pre-generate it and just send it along'. But that did not speed up anything, since `natsorted(...)`, despite getting a `key` as an argument, will just generate a new one. So my thought process was, that if adding a check in `natsorted(...)` to skip calling `key = natsort_keygen(key, alg)` if `key` was given to `natsorted(...)`, it would speed up things quite a lot (which it did after I added it).

So why did I not just use `sorted` in the first place? Well because I did not think of it to be honest. I thought I'd just use the pre-calculated key with natsorted, since it can take the key arg and it would be faster than not providing the key - which turned out to be false because it then does work that has already been done (generating the key). Or do I misunderstand the code?

**Comment by SethMMorton** [comment]
> My (well from StackOverflow) function is faster, even without regenerating the key via natsort every cycle. I meant to ask, if I makes sense for natsort to implement some test to use a faster function to do natural sorting, like in the simple example, where there are only simple strings.

This is the purpose of the `alg` keyword - the user indicates what level of complexity they need through this keyword.  The default complexity level is the smallest. If you want to submit a PR to special-case this code path that does not increase maintenance overhead I would be intrigued.

The suggestion of peeking at the user's input to decide on algorithm would be at best very challenging to implement correctly (guessing is Bad and often wrong). and at worst make things slower.

> I thought I'd just use the pre-calculated key with natsorted, since it can take the key arg and it would be faster than not providing the key - which turned out to be false because it then does work that has already been done (generating the key). Or do I misunderstand the code?

You misunderstand the purpose of a `key` function.. It is not intended to *replace* the algorithm used internal to the function, but to *transform the input data* before sending to the algorithm. This is the same as the `key` parameter for built-ins like `sorted`, `max`, and `min`.  Passing a pre-generated `key` to `natsorted` would just cause the algorithm to be run twice on the input.

The correct methodology to get what you want is in fact to use `sorted`, not `natsorted`, because in reality `natsorted` is a convenience for people that do not want to generate the key themselves ahead of time or who don't care about regenerating the key each time `natsorted` is called.

**Comment by SethMMorton** [comment]
Also, when doing your timing tests, I hope you are not using `ns.PATH`, correct?

**Comment by ganego** [comment]
> This is the purpose of the alg keyword - the user indicates what level of complexity they need through this keyword. The default complexity level is the smallest. If you want to submit a PR to special-case this code path that does not increase maintenance overhead I would be intrigued.

I did not do extensive testing on what usecases the simple function above covers correctly compared to `natsort`'s default sorting, so I cannot tell anything really sound about it unfortunately.
It of course does not handle all the special cases your library does (I know path sorting can be an issue, this is how I discovered your lib today).  
For all my usecases - basically lists of only strings (that can contain numbers) it worked fine until now.  

I'm definitely not saying defaults should be changed or this added at all as "fastStringSort" or whatever - it was meant mainly as food for thought that there is/could be quite some room for improvement, since I like optimizing stuff. Or maybe even a ns.FASTSTRINGSORT would make sense?

> Also, when doing your timing tests, I hope you are not using `ns.PATH`, correct?

Yes, only simple strings: https://gist.github.com/ganego/9cbf527aa253d2b7adb4b2ef79980e15

> Passing a pre-generated key from natsort_keygen to natsorted would just cause the algorithm to be run twice on the input.  

Indeed.

> The correct methodology to get what you want is in fact to use sorted  
 
I understand.

**Comment by SethMMorton** [comment]
I understand your surprise/frustration that `natsort` is not as fast as the "simple" algorithm you found on Stack Overflow. When I first started working on `natsort`, the algorithm was not much different from that, and then it grew over time. The complexity in the library is due to one of two things:

 1. providing the user the flexibility to choose different definitions of "natural sort" (like `ns.PATH`, for example)
 2. safeguards against "bad" data being given by the user

I strongly recommend you read the `natsort` documentation's [How It Works](https://natsort.readthedocs.io/en/master/howitworks.html) section, which goes into detail how and why most (all?) of these options or safeguards were added.

Ultimately, my primary goal/focus for `natsort` is a library that will *correctly* sort a user's input naturally (by whatever definition that user defines "naturally"). Having a performant algorithm is a secondary focus. That is not to say I do not want it to be fast - I have completely re-written the library twice over the project's lifetime in order to improve the performance. I believe that the current infrastructure (using a functional/data pipeline approach) is the as fast as I can realistically make the code without sacrificing maintainability.

But, because my primary goal is correctness, there are two steps that `natsort` performs that are in addition to what you show in the "simple" algorithm which will slow it down. Those steps are:
1. Normalize unicode input (#44)
   - Basically ensure that unicode characters that look the same but are not encoded the same all are normalized to the same encoding convention. This prevents things like `"é"` coming after `"z"` instead of after `"e"`.
2. Ensure that after going through the `natsort` algorithm, each tuple of strings follows the pattern "string, number, string, number, etc.".
  - This eliminates `TypeErrors` due Python internally trying to compare a string to a number.
  - This is commonly needed when the first character in a string is a number - a `""` needs to be inserted before that number.

Both of these slow down the processing, especially the latter.  It would be possible to omit them, but at the risk of either giving incorrectly sorted results or having the program crash.

I *could* imagine adding a mode called `ns.UNSAFE` or something that disabled these. But, after maintaining this library for so many years, I have found that unexpected data happens all the time, and encouraging users to embrace the safety at the expense of a bit of speed is probably a good thing.
